src/imitation/scripts/parallel.py

```diff
@@ -95,15 +113,47 @@ def parallel(
     )
 
     ray.init(**init_kwargs)
+    updated_tune_run_kwargs = copy.deepcopy(tune_run_kwargs)
+    if repeat > 1:
+        if "search_alg" not in updated_tune_run_kwargs:
+            updated_tune_run_kwargs["search_alg"] = optuna.OptunaSearch()
+        try:
+            algo = updated_tune_run_kwargs["search_alg"]
+            algo = search.Repeater(algo, repeat)
+            updated_tune_run_kwargs["search_alg"] = algo
+        except AttributeError:
+            raise ValueError(
+                "repeat > 1 but search_alg is not an instance of "
+                "ray.tune.search.SearchAlgorithm",
+            )
```

Comment:
Good suggestion! I often forget about the `get` method in dictionary.

> Also: why are we using Optuna if we just want to repeat some runs? This seems like overkill and a simple grid search might have less overhead?

The point of adding `repeat` is that we want to perform multiple runs of all the sampled trials. So let's say if `learning_rate` is sampled from `loguniform(1e-5, 1e-2)`, then if a value of $1.3 * 10^{-4}$ was sampled for a trial, then our `repeat=n` argument will perform this trial with the value of $1.3 * 10^{-4}$ `n` times. I believe this cannot be done using grid search unless I'm mistaken.

Regarding Optuna, I agree that we don't have to use Optuna for repeat some runs and can instead use any other search algorithm like `tune.search.basic_variant.BasicVariantGenerator` which supports both random and grid searches.

Reply:
Note from discussion with @taufeeque9:
We will stick with Optuna since the `BasicVariantGenerator` does not support being wrapped with a repeater.

Details here: https://github.com/ray-project/ray/issues/33677